### PR TITLE
Middleware: add Base firmware update check and notification

### PR DESCRIPTION
**Pull request by 0xB10C**
This adds a periodic check for firmware updates to the middleware.
If an update is available, a the middleware sends a notification.
The App can get more information about the update by calling the
IsBaseUpdateAvailable RPC, which returns a boolean indicating if
an update is available and information about the update.

The App should call `IsBaseUpdateAvailable` after the connection is
established.

### CODE REVIEW

File: middleware/cmd/middleware/main.go [hunk 1]
```diff
@@ -42,6 +43,7 @@ func main() {
 	argumentMap["bbbCmdScript"] = *bbbCmdScript
 	argumentMap["prometheusURL"] = *prometheusURL
 	argumentMap["redisPort"] = *redisPort
+	argumentMap["firmwareUpdateInfoURL"] = "https://gist.githubusercontent.com/0xB10C/a0e067b7a6e6fdcbdce2fb1f172ec7a5/raw/85136861daa70235e4a843931b123824bb6a69e5/base-icecream.json" //*firmwareUpdateInfoURL
```
Comment (0xB10C) [inline middleware/cmd/middleware/main.go]: For review and testing. Will change back to 

```suggestion
	argumentMap["firmwareUpdateInfoURL"] = firmwareUpdateInfoURL
```

before merge.

File: middleware/src/middleware.go [hunk 2]
```diff
@@ -159,9 +173,82 @@ func (middleware *Middleware) rpcLoop() {
 	}
 }
 
+// updateCheckLoop repeatedly checks for information about new Base firmware updates
+// When an update is available it's
+func (middleware *Middleware) updateCheckLoop() {
+	// This time is chosen arbitrary, but the time should be not too high for users to be notified
+	// not to long after the update release and not too low to avoid to frequent update checks.
+	const timeBetweenUpdateChecks time.Duration = 30 * time.Minute
```
Comment (0xB10C) [inline middleware/src/middleware.go]: I personally think a 30 minute update check interval is a good compromise between too frequent and to infrequent. But this is chosen arbitrarily. Looking for other suggestions.

File: middleware/src/middleware.go [hunk 3]
```diff
@@ -159,9 +173,82 @@ func (middleware *Middleware) rpcLoop() {
...
+		}
+
+		time.Sleep(timeBetweenUpdateChecks)
+	}
+}
+
+// getBaseUpdateInfo GETs a JSON file over HTTP which includes information about the update.
+func getBaseUpdateInfo(url string) (updateInfo rpcmessages.UpdateInfo, err error) {
+	client := http.Client{
+		Timeout: 15 * time.Second, // timeout is chosen arbitrary, but should account for (very) slow tor connections.
```
Comment (0xB10C) [inline middleware/src/middleware.go]: The HTTP connection timeout of 15 seconds is chosen arbitrary. It should account for very slow tor connections. Let me know what you think.

**Comment by TheCharlatan** [comment]
Concept Nit: Since the middleware will also have to handle the updates of the HSM firmware, I'd propose calling it 'Base Image' instead of 'Base Firmware'

File: middleware/src/util/semver/semver.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file is copied from the bitbox-wallet-app (https://github.com/digitalbitbox/bitbox-wallet-app/blob/master/util/semver/semver.go).
+// Modifications:
+// - change errp to errors.New()
```
Comment (0xB10C) [inline middleware/src/util/semver/semver.go]: I copy and pasted the semver.go packet from the App. I'm not to happy with it, but I think it's better to `vendor/` bigger parts of the App into the Base repository. Happy for suggestions here as well (especially from @benma).

File: middleware/src/util/semver/semver.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 Shift Devices AG
```
Comment (benma) [inline middleware/src/util/semver/semver.go]: could update this to, 2019 Shift Cryptosecurity AG

File: middleware/cmd/middleware/main.go [hunk 6]
```diff
@@ -42,6 +43,7 @@ func main() {
 	argumentMap["bbbCmdScript"] = *bbbCmdScript
 	argumentMap["prometheusURL"] = *prometheusURL
 	argumentMap["redisPort"] = *redisPort
+	argumentMap["imageUpdateInfoURL"] = "https://gist.githubusercontent.com/0xB10C/a0e067b7a6e6fdcbdce2fb1f172ec7a5/raw/85136861daa70235e4a843931b123824bb6a69e5/base-icecream.json" //*imageUpdateInfoURL
```
Comment (benma) [inline middleware/cmd/middleware/main.go]: a gist? let's use shiftcrypto.ch like for the App, or maybe github.com/digitalbitbox/bitbox-base/releases/?

File: middleware/src/middleware.go (hunk 2 above)
Comment (Stadicus) [reply middleware/src/middleware.go]: I think twice a day would suffice as well, if we could trigger a manual check from the App (e.g. if we announce an update), but that's not really important. I'm fine with just checking every 30 minutes.

**Comment by Stadicus** [comment]
> this should be changed on a repository wide (e.g. docs & shell scripts) level (@Stadicus).

Addressed in https://github.com/digitalbitbox/bitbox-base/pull/236

File: middleware/src/middleware.go (hunk 2 above)
Comment (0xB10C) [reply middleware/src/middleware.go]: Will leave it at 30 minutes.

File: middleware/src/middleware.go (hunk 3 above)
Comment (0xB10C) [reply middleware/src/middleware.go]: Going with 15 seconds here.

File: middleware/cmd/middleware/main.go (hunk 6 above)
Comment (0xB10C) [reply middleware/cmd/middleware/main.go]: Sure. The gist is for testing and review as described above.

**Comment by 0xB10C** [comment]
Opend a PR for the `base.json` file on shiftcrypto.ch https://github.com/shiftdevices/website_design/pull/210

**Comment by 0xB10C** [comment]
Rebased on to #227. https://github.com/digitalbitbox/bitbox-base/pull/235/commits/37551f98226bc90f44a1c43e603e9c9cbfcd1b50 adds the authentication stuff. @TheCharlatan please have a look if I missed anything. 
Maybe https://github.com/shiftdevices/bitbox-base-internal/issues/269 could need an update as well.

**Comment by TheCharlatan** [comment]
lgtm

**Comment by 0xB10C** [comment]
Rebased and resolved merge conflicts.

**Pull request merged**